RFC-0017-PyTorch-Operator-Versioning.md

```diff
@@ -0,0 +1,223 @@
+    * e2e FC-breaking guide
+        * It’s OK to add new optional keyword-only arguments as long as their default semantic preserve the operator’s current semantics
+
+Note that the proposal does not introduce an explicit version to _all_ PyTorch operators. Instead code changes are only required for updated operators with BC/FC breakage, that cannot be handled by automatic BC/FC methods. For other operators, the implicit version is v0.
+
+As an example, there’s a BC/FC breaking update on operator foo.
+
+Before:
+```
+foo(Tensor self, Scaler alpha=1, Tensor b, *, Tensor(a!) out) -> Tensor(a!)
+```
+After:
+```
+foo(Tensor self, Tensor c, Scaler alpha=1, Tensor b, *, Tensor(a!) out) -> Tensor(a!)
+```
+In schema, a Tensor argument, c is added. Note that it’s not added as a “tailing default argument”, so that BC/FC cannot be handled automatically.
+
+Accordingly, in the kernel of foo, the implementation is updated based on the new argument c. The pseudo code (it’s in Python format, but can be written in C++ as well) looks like:
+
+
+```python
+def foo(Tensor self, Tensor c, Scaler alpha=1, Tensor b, *, Tensor(a!) out) -> Tensor(a!):
+  # The original kernel implementation
+  ...
+  if not c.empty():
+    a.add_(c)
+```
+
+
+
+## Code changes (minimize the work of developer)
+
+If there is a BC/FC break with schema change in a PR, a lint error can be automatically generated, with instructions below to update the PR. 
+
+
+### Version bump
+
+Update version field in _operator_versions.yaml_.
+The current table that torchscript uses should be migrated to _operator_versions.yaml_.
+
+### BC updates
+
+The developer needs to implement a BC “upgrader” in Python. The upgrader code is put in a centralized python file, _operator_upgraders.py_, in TorchScript format.
+
+_operator_version.yaml_
+```yaml
+- func: foo
+ version: 10
+ upgrader: foo_upgrader_0_9
+ version: 25
+ upgrader: foo_upgrader_10_24
+```
+_operator_upgraders.py_
+```python
+def foo_upgrader_0_9(Tensor self, Tensor c, Scaler alpha=1, Tensor b, *, Tensor(a!) out):
+  c = at.empty()
+  foo(self, c, alpha, b, out = self) 
+
+def foo_upgrader_10_24(...):
+  ...
+```
+
+* Different from the upgraders defined[ here](https://github.com/pytorch/pytorch/blob/8a094e3270d2fbec6060099b7059898f4a1c104a/torch/csrc/jit/frontend/builtin_functions.cpp#L98), 
+* For some operator updates, it’s not possible to have a BC adapter. If it's FC break, an upgrader is not needed. In such a case, the operator number could still help to check compatibility and to quickly detect the source of failure with meaningful error. 
+* For most of the operator changes, the upgrader code is not expected to be heavy. However, the performance overhead should be observed, especially for edge use cases.
+* If there are multiple upgraders (for example, v20 runtime loading both v0 and v10 models). 
+* If the breakage is hard (there is no upgrader), throw “version not supported” error and suggest to regenerate the model.
```

Comment:
If recreating the model is possible we should sanity-check a BC window for mobile instead of having to build all this architecture

Reply:
Removed it since it can be covered by BC window. Inside the window no BC breakage is allowed.